crates/primitives/src/l1.rs

```diff
@@ -116,6 +116,46 @@ impl L1TxProof {
     }
 }
 
+/// Includes `L1BlockManifest` along with scan rules that it is applied to
+#[derive(Clone, Debug, PartialEq, Eq, BorshSerialize, BorshDeserialize, Arbitrary)]
+pub struct EpochedL1BlockManifest {
+    /// The actual l1 manifest
+    manifest: L1BlockManifest,
+    /// Epoch, whose scan rules are applied to this manifest
+    epoch: Epoch,
+}
+
+/// Epoch that is either exact or transitioning from one to another.
+#[derive(Clone, Debug, PartialEq, Eq, BorshSerialize, BorshDeserialize, Arbitrary)]
+pub enum Epoch {
+    /// Exact epoch
+    Exact(u64),
```

Comment:
Probably makes more sense to say "within" instead of "exact"?

Should also have some accessor that takes away the nuance this is trying to add for when we just need a specific epoch number.  The name `Epoch` here seems too broad.  Really the epoch index should always be represented as a single number.

Reply:
Yeah the name `Epoch` is too broad in this context. It's true that epoch should mean a single number, but in order to handle valid transactions that can fall under both old and new epoch, such a construct was used.